src/tokens/blockchain.service.ts

```diff
@@ -88,15 +104,60 @@ export class BlockchainConnectorService {
     });
   }
 
+  // Check if retry condition matches the err that's been hit
+  private matchesRetryCondition(err: any): boolean {
+    return this.retryCondition != '' && err?.toString().match(this.retryCondition) !== null;
+  }
+
+  // Delay by the appropriate amount of time given the iteration the caller is in
+  private async backoffDelay(iteration: number) {
+    const delay = Math.min(
+      this.retryBackOffInitial * Math.pow(this.retryBackOffFactor, iteration),
+      this.retryBackOffLimit,
+    );
+    await new Promise(resolve => setTimeout(resolve, delay));
+  }
+
+  // Generic helper function that makes an given blockchain function retryable
+  // by using synchronous, back-off delays for cases where the function returns
+  // an error which matches the configured retry condition
+  private async retryableCall<T = any>(
+    blockchainFunction: () => Promise<AxiosResponse<T>>,
+  ): Promise<AxiosResponse<T>> {
+    let response: any;
+    for (let retries = 0; retries <= this.retriesMax; retries++) {
+      try {
+        response = await blockchainFunction();
+        break;
+      } catch (e) {
+        if (this.matchesRetryCondition(e)) {
+          this.logger.debug(`Retry condition matched for error ${e}`);
+          // Wait for a backed-off delay before trying again
+          await this.backoffDelay(retries);
+        } else {
+          // Whatever the error was it's not one we will retry for
+          break;
+        }
+      }
+    }
+
+    return response;
+  }
+
   async query(ctx: Context, to: string, method?: IAbiMethod, params?: any[]) {
-    const response = await this.wrapError(
-      lastValueFrom(
-        this.http.post<EthConnectReturn>(
-          this.baseUrl,
-          { headers: { type: queryHeader }, to, method, params },
-          this.requestOptions(ctx),
-        ),
-      ),
+    const url = this.baseUrl;
+    const response = await this.retryableCall<EthConnectReturn>(
+      async (): Promise<AxiosResponse<EthConnectReturn>> => {
```

Comment:
Not sure I see the need to wrap this again - can't you just pass the return of `this.wrapError(...)` directly to `this.retryableCall`?

Reply:
It also feels a little weird that we go to the trouble of catching and wrapping the Axios error in `wrapError`, and _then_ pass the wrapped error to `retryableCall` for its regex processing. I think in most cases we'll end up catching the Axios error, wrapping it in `InternalServerErrorException`, stringifying that wrapped result, then parsing it against the regex. It probably works OK, but it feels just slightly wonky.

Reply:
Yeah I initially went with the approach of "make the entire current call stack retryable" so I left everything from `wrapError()` down to be part of the retryable function. But wrapping the retryable code does make more sense I think so the latest commit uses that approach.